web/packages/teleterm/src/ui/DocumentGateway/useGateway.ts

```diff
@@ -39,11 +39,14 @@ export function useGateway(doc: types.DocumentGateway) {
   // input to a controlled one once `doc.port` gets set. The backend will handle converting an empty
   // string to '0'.
   const defaultPort = doc.port || '';
-  const gateway = ctx.clustersService.findGateway(doc.gatewayUri);
+  const gateway = useStoreSelector(
+    'clustersService',
+    useCallback(state => state.gateways.get(doc.gatewayUri), [doc.gatewayUri])
+  );
```

Comment:
Previously, `DocumentGatewayApp` would call `ctx.clustersService.useState()` to know about gateway changes. `DocumentGateway` did no such thing. I suspect this all worked only because when you call `createGateway`, it updates `documentsService`, which in turns re-renders all components that use `useWorkspaceContext`, but also this would result in `doc` from props being changed.

Now `useGateway` is subscribed to updates of this specific gateway. From what I've seen, this doesn't cause any issues since this is the only piece of state in this hook, other than `doc` which is updated through `documentsService` and passed as a prop. 

---

As an aside, components like `DocumentGateway` do not require being updated each time workspace state is updated. This is because they don't read anything from `documentsService`, they call `useWorkspaceContext` merely so that their event handlers can add new documents to the workspace.

However, since we use class services in many places and there's no meaningful barrier between functions that read data and functions that write data, `useWorkspaceContext` must assume that the caller can read the workspace state.